### PR TITLE
Test utf8 validaiton based on rust-isutf8

### PR DESCRIPTION
**Pull request by Licenser**
# DO NOT MERGE

This is a copy of @ArniDagur's rust-isutf8 port of Lemire's utf8 validation to use for validation to test the performance difference as suggested in #68455

### CODE REVIEW

**Comment by Licenser** [comment]
cc @yoshuawuyts

File: src/libcore/str/mod.rs [hunk 1]
```diff
@@ -1484,19 +1485,452 @@ impl<'a> DoubleEndedIterator for LinesAny<'a> {
 #[allow(deprecated)]
 impl FusedIterator for LinesAny<'_> {}
 
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+mod avx2 {
```
Comment (CryZe) [inline src/libcore/str/mod.rs]: Shouldn't this specify a bunch of target features? If you don't do so it generates MUCH worse code (and maybe it doesn't even work, not sure).

Comment (Licenser) [reply src/libcore/str/mod.rs]: @CryZe good catch, I updated it the guard to use `target_feature = "avx2"` that should solve this

**Comment by ArniDagur** [comment]
Is that 44% _slower_? Can that be right?

**Comment by CryZe** [comment]
The code was without the AVX2 target feature guard, which has a heavy performance penalty. The timer needs to be rerun.

**Comment by mati865** [comment]
`x86_64` rightfully doesn't have `target_feature = "avx2"` by default.
Checks like `#[cfg(any(target_feature = "avx2"))]` will just return false.

Simply enabling AVX or changing CPU to `znver2` (that's what perf.rlo is running on) won't give meaningful results either because base toolchain would also need to be built with that.

**Comment by CryZe** [comment]
Yes, iiuc you need to runtime select between the simd and non-simd versions.

**Comment by Mark-Simulacrum** [comment]
I don't quite follow the discussion. If we need another timer run, please say so explicitly :)

**Comment by yoshuawuyts** [comment]
What's interesting to me is how much of a *negative* impact utf-8 validation seems to have. Like @Mark-Simulacrum postulated it indeed affects shorter programs more than longer ones (as evidenced by "hello world"). This makes it interesting to consider to what degree a *positive* change to the algorithm might improve things.

---
@Mark-Simulacrum https://github.com/rust-lang/rust/pull/68463/commits/1a248181a15998866c3ba6eccd4b743bd36fc499 includes fixes that should speed things up; another run would be great.

**Comment by CryZe** [comment]
@yoshuawuyts I don't think that commit is sufficient. rustc is not built with avx2 active, so it doesn't build the code. It needs some code to select which version of the algorithm to use at runtime.

**Comment by ArniDagur** [comment]
Is there no way to test it with `-C target-cpu=skylake` or equivelent, just for this PR?

**Comment by Mark-Simulacrum** [comment]
I do not have the time to suggest a way to enable those compile time flags here (especially because our CI is not guaranteed to run on skylake). If you do not want to dig in, then I would propose that we close this.

**Comment by Licenser** [comment]
I don’t have any control about the benchmark system it’s not about want i can’t change that.

**Pull request closed without merging**